src/wifi_dpp.rs

```diff
@@ -0,0 +1,451 @@
+                    _phantom: PhantomData,
+                })
+            }
+            DppState::Fail(e) => Err(e),
+            other => Err(unexpected_state(other)),
+        }
+    }
+
+    fn do_bootstrap_gen(
+        channels: &[u8],
+        key: Option<&[u8; 32]>,
+        associated_data: Option<&[u8]>,
+    ) -> Result<(), EspError> {
+        let mut channels_str = channels.into_iter().fold(String::new(), |mut a, c| {
+            write!(a, "{c},").unwrap();
+            a
+        });
+        channels_str.pop();
+        let channels_cstr = CString::new(channels_str).unwrap();
+
+        let key_ascii_cstr = key.map(|k| {
+            let result = frame_key(k).iter().fold(String::new(), |mut a, b| {
+                write!(a, "{b:02x}").unwrap();
+                a
+            });
+            CString::new(result).unwrap()
+        });
+
+        let associated_data_cstr = match associated_data {
+            Some(associated_data) => Some(CString::new(associated_data).map_err(|_| {
+                warn!("associated data contains an embedded NUL character!");
+                EspError::from_infallible::<ESP_ERR_INVALID_ARG>()
+            })?),
+            None => None,
+        };
+
+        info!("Bootstrapping DPP with: channels={channels_cstr:?}, key={key_ascii_cstr:?}");
+        esp!(unsafe {
+            esp_supp_dpp_bootstrap_gen(
+                channels_cstr.as_ptr(),
+                dpp_bootstrap_type_DPP_BOOTSTRAP_QR_CODE,
+                key_ascii_cstr
+                    .as_ref()
+                    .map_or_else(ptr::null, |x| x.as_ptr()),
+                associated_data_cstr
+                    .as_ref()
+                    .map_or_else(ptr::null, |x| x.as_ptr()),
+            )
+        })?;
+
+        // Guarantees we get a compiler error if we mess up the lifetime...
+        drop(channels_cstr);
+        drop(key_ascii_cstr);
+        drop(associated_data_cstr);
+
+        Ok(())
+    }
+}
+
+impl<'d, 'w, T> EspWifiDpp<'d, 'w, T> {
+    pub fn get_bootstrapped_data(&self) -> &T {
+        &self.bootstrapped_data
+    }
+
+    pub fn start_listen(self) -> Result<EspWifiDppListener<'d, 'w, T>, EspError> {
+        EspWifiDppListener::start_listen(self)
+    }
+}
+
+pub struct EspWifiDppListener<'d, 'w, T> {
+    bootstrapped: EspWifiDpp<'d, 'w, T>,
+}
+
+impl<'d, 'w, T> EspWifiDppListener<'d, 'w, T> {
+    fn start_listen(bootstrapped: EspWifiDpp<'d, 'w, T>) -> Result<Self, EspError> {
+        info!("Starting DPP listener...");
+        esp!(unsafe { esp_supp_dpp_start_listen() })?;
+        Ok(Self { bootstrapped })
+    }
+
+    /// Blocking wait for credentials or a possibly retryable error.  Note that user error
+    /// such as scanning the wrong QR code can trigger this error case.  Retries are highly
+    /// recommended, and especially via [Self::attempt_retry].
+    pub fn wait_for_credentials(&self) -> Result<ClientConfiguration, EspError> {
```

Comment:
I hesitated on this being &mut self or &self.  It will break stuff if you call this in parallel (should hang forever), but if it's &mut self we can't expose stop_listen on Self.  Would it be preferred to split the listener into two structs, one that can be used to cancel and one that can be used to wait with some shared lock between them?

Reply:
I'm fine with the current little cheat for now. We can fix/complicate in a subsequent version.